symbolic-base/src/ZkFold/Symbolic/Data/JWT/Twitch.hs

```diff
@@ -0,0 +1,138 @@
+    PubSubPerms
+        { pListen :: VarByteString 2048 ctx
+        , pSend   :: VarByteString 2048 ctx
+        }
+    deriving Generic
+
+deriving instance
+    ( P.Eq (ctx (V.Vector 2048))
+    , P.Eq (ctx Par1)
+    ) => P.Eq (PubSubPerms ctx)
+deriving instance
+    ( P.Show (ctx (V.Vector 2048))
+    , P.Show (ctx Par1)
+    ) => P.Show (PubSubPerms ctx)
+deriving instance Symbolic ctx => SymbolicData (PubSubPerms ctx)
+deriving instance Symbolic ctx => SymbolicInput (PubSubPerms ctx)
+instance Symbolic ctx => Arbitrary (PubSubPerms ctx) where
+    arbitrary = genericArbitrary uniform
+
+instance (Symbolic ctx, Context (PubSubPerms ctx) ~ ctx) => IsSymbolicJSON (PubSubPerms ctx) where
+    type MaxLength (PubSubPerms ctx) = 4248
+    toJsonBits PubSubPerms{..} =
+                    (fromType @"{\"listen\":") @+ pListen
+        `VB.append` (fromType @",\"send\":")   @+ pSend
+        `VB.append` (fromType @"}")
+
+instance Symbolic ctx => FromJSON (PubSubPerms ctx) where
+    parseJSON = genericParseJSON (aesonPrefix snakeCase) . stringify
+
+deriving instance
+    ( P.Eq (ctx (V.Vector 40))
+    , P.Eq (ctx (V.Vector 80))
+    , P.Eq (ctx (V.Vector 88))
+    , P.Eq (ctx (V.Vector 512))
+    , P.Eq (ctx (V.Vector 2048))
+    , P.Eq (ctx Par1)
+    ) => P.Eq (TwitchPayload ctx)
+
+deriving instance
+    ( P.Show (ctx (V.Vector 40))
+    , P.Show (ctx (V.Vector 80))
+    , P.Show (ctx (V.Vector 88))
+    , P.Show (ctx (V.Vector 512))
+    , P.Show (ctx (V.Vector 2048))
+    , P.Show (ctx Par1)
+    ) => P.Show (TwitchPayload ctx)
+deriving instance Symbolic ctx => SymbolicData (TwitchPayload ctx)
+deriving instance Symbolic ctx => SymbolicInput (TwitchPayload ctx)
+instance Symbolic ctx => Arbitrary (TwitchPayload ctx) where
+    arbitrary = genericArbitrary uniform
+
+instance Symbolic ctx => FromJSON (TwitchPayload ctx) where
+    parseJSON = genericParseJSON (aesonPrefix snakeCase) . stringify
+
+instance (Symbolic ctx, Context (TwitchPayload ctx) ~ ctx) => IsSymbolicJSON (TwitchPayload ctx) where
+    type MaxLength (TwitchPayload ctx) = 6792
+    toJsonBits TwitchPayload{..} =
+                    (fromType @"{\"channel_id\":\"")   @+ twChannelId
+        `VB.append` (fromType @"\",\"exp\":\"") @+ twExp
+        `VB.append` (fromType @"\",\"is_unlinked\":") @+ twIsUnlinked
+        `VB.append` (fromType @",\"opaque_user_id\":\"") @+ twOpaqueUserId
+        `VB.append` (fromType @"\",\"pubsub_perms\":")  @+ toJsonBits twPubsubPerms
+        `VB.append` (fromType @",\"role\":\"") @+ twRole
+        `VB.append` (fromType @"\",\"user_id\":\"") @+ twUserId
+        `VB.append` (fromType @"\"}")
+
+instance
+  ( Symbolic ctx
+  , NFData (ctx (V.Vector 8))
+  , NFData (ctx (V.Vector 9056))
+  ) => IsBits (TwitchPayload ctx) where
+    type BitCount (TwitchPayload ctx) = 9056
+    toBits = toAsciiBits
+
+instance (Symbolic ctx, TokenBits (TwitchPayload ctx), RSA.RSA 2048 10328 ctx) => IsTokenPayload "RS256" (TwitchPayload ctx) where
```

Comment:
Why is the `TokenBits (TwitchPayload ctx)` constraint needed here?

Reply:
because both `signPayload` and `verifyJWT` use `tokenBits` from this class

Reply:
Oh, ok, I mean shouldn't this constraint be inferrable by itself? Or is it that the needed instance is currently missing?

Reply:
Oh, no, TokenBits also includes several NFData and KnownNat constraints. If you remove it, GHC complains that
```src/ZkFold/Symbolic/Data/JWT/Twitch.hs:137:56: error: [GHC-39999]
    • Could not deduce ‘NFData (ctx (V.Vector 8))’
        arising from a use of ‘tokenBits’
```

Reply:
Oh, ok, understood!